clients/da-vinci-client/src/test/java/com/linkedin/davinci/validation/TestPartitionTracker.java

```diff
@@ -438,4 +498,181 @@ public void testMidSegmentCheckSumStates(boolean isVersionTopic, CheckSumType ch
+         *       we're looking for.
+         */
+        assertFalse(e instanceof ImproperlyStartedSegmentException);
+      }
+
+      /** But if we do consume the correct message (monotonically speaking), then it's fine. */
+      this.partitionTracker.validateMessage(type, currentMessage, true, Lazy.FALSE);
+    }
+
+    assertTrue(segmentId > 0);
+  }
+
+  private List<DefaultPubSubMessage> getSegmentOfMessages(
+      TopicType type,
+      CheckSumType checkSumType,
+      int segmentId,
+      long startingOffset,
+      ByteBuffer eosChecksumValue) {
+    PubSubTopicPartition pubSubTopicPartition = getPubSubTopicPartition(type);
+    Segment segment = new Segment(partitionId, segmentId, checkSumType);
+    long offset = startingOffset;
+
+    List<DefaultPubSubMessage> messages = new ArrayList<>();
+
+    // 1st message: SOS
+    ControlMessage startOfSegment = getStartOfSegment(checkSumType);
+    KafkaMessageEnvelope sosMessage =
+        getKafkaMessageEnvelope(MessageType.CONTROL_MESSAGE, guid, segment, Optional.empty(), startOfSegment);
+    DefaultPubSubMessage sosConsumerRecord = new ImmutablePubSubMessage(
+        getControlMessageKey(sosMessage),
+        sosMessage,
+        pubSubTopicPartition,
+        ApacheKafkaOffsetPosition.of(offset++),
+        System.currentTimeMillis() + 1000,
+        0);
+    messages.add(sosConsumerRecord);
+
+    // 2nd message: PUT
+    Put firstPut = getPutMessage("first_message".getBytes());
+    KafkaMessageEnvelope putMessage =
+        getKafkaMessageEnvelope(MessageType.PUT, guid, segment, Optional.empty(), firstPut);
+    KafkaKey firstMessageKey = getPutMessageKey("first_key".getBytes());
+    DefaultPubSubMessage putConsumerRecord = new ImmutablePubSubMessage(
+        firstMessageKey,
+        putMessage,
+        pubSubTopicPartition,
+        ApacheKafkaOffsetPosition.of(offset++),
+        System.currentTimeMillis() + 1000,
+        0);
+    messages.add(putConsumerRecord);
+
+    // 3rd message: UPDATE
+    Update firstUpdate = getUpdateMessage("second_message".getBytes());
+    KafkaMessageEnvelope updateMessage =
+        getKafkaMessageEnvelope(MessageType.UPDATE, guid, segment, Optional.empty(), firstUpdate);
+    KafkaKey secondMessageKey = getUpdateMessageKey("second_key".getBytes());
+    DefaultPubSubMessage updateConsumerRecord = new ImmutablePubSubMessage(
+        secondMessageKey,
+        updateMessage,
+        pubSubTopicPartition,
+        ApacheKafkaOffsetPosition.of(offset++),
+        System.currentTimeMillis() + 1000,
+        0);
+    messages.add(updateConsumerRecord);
+
+    // 4th message: DELETE
+    Delete firstDelete = new Delete();
+    KafkaMessageEnvelope deleteMessage =
+        getKafkaMessageEnvelope(MessageType.DELETE, guid, segment, Optional.empty(), firstDelete);
+    KafkaKey thirdMessageKey = getDeleteMessageKey("third_key".getBytes());
+    DefaultPubSubMessage deleteConsumerRecord = new ImmutablePubSubMessage(
+        firstMessageKey,
```

Comment:
```suggestion
        thirdMessageKey,
```
Caught by SpotBugs. 🐞